src/System.Linq.Expressions/tests/BinaryOperators/Assignment/Assign.cs

```diff
@@ -187,24 +256,92 @@ public void ReferenceAssignable(bool useInterpreter)
             Assert.Equal("Hello", Expression.Lambda<Func<object>>(exp).Compile(useInterpreter)());
         }
 
-        [Theory, MemberData(nameof(ReadOnlyExpressions))]
-        public void AttemptToAssignToNonWritable(Expression readonlyExp)
+        [Theory]
+        [MemberData(nameof(ReadOnlyExpressions))]
+        public void LeftReadOnly_ThrowsArgumentException(Expression readonlyExp)
         {
             Assert.Throws<ArgumentException>("left", () => Expression.Assign(readonlyExp, Expression.Default(readonlyExp.Type)));
         }
 
-        [Theory, MemberData(nameof(WriteOnlyExpressions))]
-        public static void AttemptToAssignFromNonReadable(Expression writeOnlyExp)
+        [Theory]
+        [MemberData(nameof(WriteOnlyExpressions))]
+        public static void Right_WriteOnly_ThrowsArgumentException(Expression writeOnlyExp)
         {
             ParameterExpression variable = Expression.Variable(writeOnlyExp.Type);
             Assert.Throws<ArgumentException>("right", () => Expression.Assign(variable, writeOnlyExp));
         }
 
+        [Theory]
+        [ClassData(typeof(InvalidTypes))]
+        public static void Left_InvalidType_ThrowsArgumentException(Type type)
+        {
+            Expression left = new FakeExpression(ExpressionType.Parameter, type);
+            Assert.Throws<ArgumentException>("left", () => Expression.Assign(left, Expression.Parameter(typeof(int))));
+        }
+
+        [Theory]
+        [ClassData(typeof(InvalidTypes))]
+        public static void Right_InvalidType_ThrowsArgumentException(Type type)
+        {
+            Expression right = new FakeExpression(ExpressionType.Parameter, type);
+            Assert.Throws<ArgumentException>("right", () => Expression.Assign(Expression.Variable(typeof(object)), right));
+        }
+
+        [Theory]
+        [ClassData(typeof(CompilationTypes))]
+        public static void Left_ValueTypeContainsChildTryExpression_ThrowsNotSupportedExceptionOnCompilation(bool useInterpreter)
+        {
+            Expression tryExpression = Expression.TryFinally(
+                Expression.Constant(1),
+                Expression.Empty()
+            );
+            Expression index = Expression.Property(Expression.Constant(new StructWithPropertiesAndFields()), typeof(StructWithPropertiesAndFields).GetProperty("Item"), new Expression[] { tryExpression });
+
+            Expression<Func<bool>> func = Expression.Lambda<Func<bool>>(
+                Expression.Block(
+                    Expression.Assign(index, Expression.Constant(123)),
+                    Expression.Equal(index, Expression.Constant(123))
+                    )
+                );
+
+            if (useInterpreter)
+            {
+                Assert.True(func.Compile(useInterpreter)());
+            }
+            else
+            {
+                Assert.Throws<NotSupportedException>(() => func.Compile(useInterpreter));
```

Comment:
Yeah this is an unfortunate restriction in the compiler due to stack spilling because it occurs as an expression rewrite step prior to `LambdaCompiler` and `ILGen` and we don't have an object model for obtaining the `ref` of a value nor storing it in a `ref` local. Hence we have to throw rather than performing the stack spilling which would result in creating a *copy* of the value rather than mutating the value through a reference to it. At some point, we could look into trying to remove this restriction, either by supporting `ref` locals possibly only in an internal-only variation of a block node (just like we have `SpilledExpressionBlock`) or by moving the spilling to a later phase in the compiler.

FWIW, I've been writing some tests for various stack spilling cases a while back. I'll send out a PR once the list is a bit more complete.